README
AnyEvent-RipeRedis version 0.34
===============================

INSTALLATION

To install this module type the following:

  perl Makefile.PL
  make
  make test
  make install

DESCRIPTION

AnyEvent::RipeRedis is flexible non-blocking Redis client. Supports
subscriptions, transactions and can automaticaly restore connection after
failure.

Requires Redis 1.2 or higher and any supported event loop.

AUTHOR

Eugene Ponizovsky, <ponizovsky@gmail.com>

Sponsored by SMS Online, <dev.opensource@sms-online.com>

Special thanks

* Alexey Shrub
* Vadim Vlasov
* Konstantin Uvarin
* Ivan Kruglov

COPYRIGHT AND LICENSE

Copyright (c) 2012-2017, Eugene Ponizovsky, SMS Online. All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.


Changes
Revision history for AnyEvent::RipeRedis.

0.44 Fri Dec 1 12:09:02 MSK 2017
  - Made that the client do not reconnects on the error
    "Client sent AUTH, but no password is set".

0.42 Sat Apr 1 15:49:15 MSK 2017
  - README.pod replaced by README.md.

0.40 Sat Feb 25 16:54:34 MSK 2017
  - Fix in Changes file.

0.38 Sat Feb 25 16:38:55 MSK 2017
  - Added error code E_NOT_BUSY.
  - Minor fixes.

0.36 Tue Feb 21 14:55:42 MSK 2017
  - Added parsing of reply for "CLUSTER INFO" command.
  - Added README.pod file instead of README file.

0.34 Sat Jan 28 18:59:15 MSK 2017
  - Light refactoring.

0.32 Thu Dec 29 15:16:05 MSK 2016
  - Weakened one more circular reference.
  - Fixed typo in POD.

0.30 Sun Nov 20 19:40:55 MSK 2016
  - Fixes in POD.
  - Fixes in tests.
  - Cosmetic changes in code.

0.28 Fri Sep 23 12:10:33 MSK 2016
  - Bugfix. If the client object was destroyed prematurely, in some cases
    "_process_reply" method have been called on undefined value.

0.26 Fri Sep 2 18:17:17 MSK 2016
  - "min_reconnect_interval" renamed to "reconnect_interval"

0.24 Fri Sep 2 08:20:05 MSK 2016
  - Error "not allowed after "multi"" now raise exception.
  - Unit tests improved.

0.22 Tue Aug 30 19:51:25 MSK 2016
  - Unit tests improved.

0.20 Tue Aug 30 19:20:30 MSK 2016
  - Minor changes in POD.
  - Cosmetic changes in tests.

0.18 Tue Aug 30 11:03:25 MSK 2016
  - Added link to AnyEvent::RipeRedis::Cluster in POD.

0.16 Tue Aug 30 11:00:00 MSK 2016
  - %ERROR_CODES moved to AnyEvent::RipeRedis::Error.
  - Minor changes in unit tests.
  - Changes in POD.

0.14 Mon Aug 29 17:01:40 MSK 2016
  - Removed unnecessary string in unit test.

0.12 Mon Aug 29 16:48:05 MSK 2016
  - Refactoring of unit test.

0.10 Wed Aug 26 11:44:35 MSK 2016
  - Removed "on_connect_error" callback.
  - POD Improved.
  - Unit tests improved.

0.08 Wed Aug 25 11:18:30 MSK 2016
  - Added method "execute()".
  - Getter "selected_database()" renamed to "database()".
  - Bugfix. "DISCARD" command do not turned off transaction mode.
  - Refactoring.
  - POD Improved.

0.06 Fri Jul 29 10:51:10 MSK 2016
  - Improved POD.
  - Light refactoring.

0.04 Mon Jul 27 11:41:15 MSK 2016
  - Refactoring of eval_cached() method.
  - POD improved.
  - Bugfix in unit tests.
  - Removed unused code. Cosmetic.

0.02 Mon Jul 25 15:37:30 MSK 2016
  - First release of new incarnation of AnyEvent::Redis::RipeRedis.


Makefile.PL
use 5.008000;

use ExtUtils::MakeMaker;

WriteMakefile(
  NAME             => 'AnyEvent::RipeRedis',
  VERSION_FROM     => 'lib/AnyEvent/RipeRedis.pm',
  MIN_PERL_VERSION => '5.008000',
  PREREQ_PM        => {
    'AnyEvent'     => '6.01',
    'Encode'       => '0',
    'Scalar::Util' => '0',
    'Digest::SHA'  => '0',
    'Carp'         => '0',
  },
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '6.64',
  },
  BUILD_REQUIRES => {
    'Test::More'  => '0.98',
    'Test::TCP'   => '2.12',
    'Test::Fatal' => '0.013',
    'Time::HiRes' => '0',
    'File::Temp'  => '0.19',
    'POSIX'       => '0',
    version       => '0.77',
  },
  META_MERGE => {
    resources => {
      homepage   => 'https://github.com/iph0/AnyEvent-RipeRedis',
      bugtracker => 'https://github.com/iph0/AnyEvent-RipeRedis/issues',
      repository => 'https://github.com/iph0/AnyEvent-RipeRedis',
      license    => 'http://dev.perl.org/licenses/',
    },
  },
  ABSTRACT_FROM => 'lib/AnyEvent/RipeRedis.pm',
  AUTHOR        => 'Eugene Ponizovsky <ponizovsky@gmail.com>',
  LICENSE       => 'perl',
);


examples/subs.pl
#!/usr/bin/perl

use strict;
use warnings;

use AnyEvent;
use AnyEvent::RipeRedis;

my $cv = AE::cv;

my $redis = AnyEvent::RipeRedis->new(
  host     => 'localhost',
  port     => 6379,
  password => 'redis_pass',

  on_connect => sub {
    print "Connected to Redis server\n";
  },

  on_disconnect => sub {
    print "Disconnected from Redis server\n";
  },
);

# Subscribe to channels by name
$redis->subscribe( qw( foo bar ),
  { on_reply => sub {
      my $reply = shift;
      my $err   = shift;

      if ( defined $err ) {
        warn $err->message . "\n";
        return;
      };

      print "Subscribed on: foo, bar\n";
    },

    on_message => sub {
      my $message = shift;
      my $channel = shift;

      print "$channel: $message\n";
    },
  }
);

# Subscribe to channels by pattern
$redis->psubscribe( qw( info_* err_* ),
  { on_reply => sub {
      my $reply = shift;
      my $err   = shift;

      if ( defined $err ) {
        warn $err->message . "\n";
        return;
      };

      print "Subscribed on: info_*, err_*\n";
    },

    on_message => sub {
      my $message = shift;
      my $pattern = shift;
      my $channel = shift;

      print "$pattern ($channel): $message\n";
    },
  }
);

# Unsubscribe
my $on_signal = sub {
  print "Stopped\n";

  $redis->unsubscribe( qw( foo bar ),
    sub {
      my $msg = shift;
      my $err = shift;

      if ( defined $err ) {
        warn $err->message . "\n";
        return;
      };

      print "Unsubscribed from: foo, bar\n";
    }
  );

  $redis->punsubscribe( qw( info_* err_* ),
    sub {
      my $msg = shift;
      my $err = shift;

      if ( defined $err ) {
        warn $err->message . "\n";
        return;
      };

      print "Unsubscribed from: info_*, err_*\n";

      $cv->send;
    }
  );

  my $timer;
  $timer = AE::timer( 5, 0,
    sub {
      undef( $timer );
      exit 0; # Emergency exit
    },
  );
};

my $int_w  = AE::signal( INT  => $on_signal );
my $term_w = AE::signal( TERM => $on_signal );

$cv->recv;

$redis->disconnect;


examples/generic.pl
#!/usr/bin/perl

use strict;
use warnings;

use AnyEvent;
use AnyEvent::RipeRedis;

my $cv = AE::cv;

my $redis = AnyEvent::RipeRedis->new(
  host     => 'localhost',
  port     => 6379,
  password => 'redis_pass',

  on_connect => sub {
    print "Connected to Redis server\n";
  },

  on_disconnect => sub {
    print "Disconnected from Redis server\n";
  },
);

# Increment
$redis->incr( 'foo',
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
    }

    print "$reply\n";
  }
);

# Set value
$redis->set( 'bar', 'Some string',
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    print "SET is done\n";
  }
);

# Get value
$redis->get( 'bar',
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    print "$reply\n";
  }
);

# Push values
for ( my $i = 1; $i <= 3; $i++ ) {
  $redis->rpush( 'list', "element_$i",
    sub {
      my $reply = shift;
      my $err   = shift;

      if ( defined $err ) {
        warn $err->message . "\n";
        return;
      }

      print "$reply\n";
    }
  );
}

# Get list of values
$redis->lrange( 'list', 0, -1,
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    foreach my $value ( @{$reply} ) {
      print "$value\n";
    }
  }
);

# Transaction
$redis->multi(
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    print "Transaction begun\n";
  }
);
$redis->incr( 'foo',
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    print "$reply\n";
  }
);
$redis->lrange( 'list', 0, -1,
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    print "$reply\n";
  }
);
$redis->get( 'bar',
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    print "$reply\n";
  }
);
$redis->exec(
  sub {
    my $replies = shift;
    my $err     = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    foreach my $reply ( @{$replies} ) {
      if ( ref( $reply ) eq 'ARRAY' ) {
        foreach my $value ( @{$reply} ) {
          print "$value\n";
        }
      }
      else {
        print "$reply\n";
      }
    }
  }
);

# Delete keys
$redis->del( qw( foo bar list ),
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    print "$reply keys removed\n";
  }
);

# Disconnect
$redis->quit(
  sub {
    my $reply = shift;
    my $err   = shift;

    if ( defined $err ) {
      warn $err->message . "\n";
      return;
    }

    $cv->send;
  }
);

$cv->recv;


examples/unix_socket.pl
#!/usr/bin/perl

use strict;
use warnings;

use AnyEvent;
use AnyEvent::RipeRedis;

my $cv = AE::cv;

my $redis;
$redis = AnyEvent::RipeRedis->new(
  host                   => 'unix/',
  port                   => '/var/run/redis/redis.sock',
  password               => 'redis_pass',
  connection_timeout     => 5,
  read_timeout           => 5,
  min_reconnect_interval => 5,

  on_connect => sub {
    print "Connected to Redis server\n";
  },

  on_disconnect => sub {
    print "Disconnected from Redis server\n";
  },
);

my $timer;
$timer = AE::timer( 0, 1,
  sub {
    $redis->incr( 'foo',
      sub {
        my $reply = shift;
        my $err   = shift;

        if ( defined $err ) {
          warn $err->message . "\n";
          return;
        }

        print "$reply\n";
      },
    );
  },
);

my $on_signal = sub {
  print "Stopped\n";
  $cv->send;
};

my $int_w  = AE::signal( INT  => $on_signal );
my $term_w = AE::signal( TERM => $on_signal );

$cv->recv;

$redis->disconnect;


lib/AnyEvent/RipeRedis/Error.pm
package AnyEvent::RipeRedis::Error;

use 5.008000;
use strict;
use warnings;

our $VERSION = '0.44';

our %ERROR_CODES = (
  E_CANT_CONN                  => 1,
  E_LOADING_DATASET            => 2,
  E_IO                         => 3,
  E_CONN_CLOSED_BY_REMOTE_HOST => 4,
  E_CONN_CLOSED_BY_CLIENT      => 5,
  E_NO_CONN                    => 6,
  E_OPRN_ERROR                 => 9,
  E_UNEXPECTED_DATA            => 10,
  E_NO_SCRIPT                  => 11,
  E_READ_TIMEDOUT              => 12,
  E_BUSY                       => 13,
  E_MASTER_DOWN                => 14,
  E_MISCONF                    => 15,
  E_READONLY                   => 16,
  E_OOM                        => 17,
  E_EXEC_ABORT                 => 18,
  E_NO_AUTH                    => 19,
  E_WRONG_TYPE                 => 20,
  E_NO_REPLICAS                => 21,
  E_BUSY_KEY                   => 22,
  E_CROSS_SLOT                 => 23,
  E_TRY_AGAIN                  => 24,
  E_ASK                        => 25,
  E_MOVED                      => 26,
  E_CLUSTER_DOWN               => 27,
  E_NOT_BUSY                   => 28,
);


sub new {
  my $class    = shift;
  my $err_msg  = shift;
  my $err_code = shift;

  my $self = bless {}, $class;

  $self->{message} = $err_msg;
  $self->{code}    = $err_code;

  return $self;
}

sub message {
  my $self = shift;
  return $self->{message};
}

sub code {
  my $self = shift;
  return $self->{code};
}

1;
__END__

=head1 NAME

AnyEvent::RipeRedis::Error - Class of error for AnyEvent::RipeRedis

=head1 DESCRIPTION

Class of error for L<AnyEvent::RipeRedis>. Objects of this class can be passed
to callbacks.

=head1 CONSTRUCTOR

=head2 new( $err_msg, $err_code )

Creates error object.

=head1 METHODS

=head2 message()

Get error message.

=head2 code()

Get error code.

=head1 SEE ALSO

L<AnyEvent::RipeRedis>

=head1 AUTHOR

Eugene Ponizovsky, E<lt>ponizovsky@gmail.comE<gt>

Sponsored by SMS Online, E<lt>dev.opensource@sms-online.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2012-2017, Eugene Ponizovsky, SMS Online. All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
